src/htm/algorithms/Connections.cpp

```diff
@@ -534,15 +534,72 @@ void Connections::raisePermanencesToThreshold(
   if( increment <= 0 ) // If minPermSynPtr is already connected then ...
     return;            // Enough synapses are already connected.
 
-  // Raise the permance of all synapses in the potential pool uniformly.
-  for( const auto &syn : synapses ) //TODO vectorize: vector + const to all members
-    updateSynapsePermanence(syn, synapses_[syn].permanence + increment); //this is performance HOTSPOT
+  // Raise the permanence of all synapses in the potential pool uniformly.
+  bumpSegment(segment, increment);
+}
+
+
+void Connections::synapseCompetition(
+                    const Segment    segment,
+                    const SynapseIdx minimumSynapses,
+                    const SynapseIdx maximumSynapses)
+{
+  NTA_ASSERT( minimumSynapses <= maximumSynapses);
+
+  const auto &segData = dataForSegment( segment );
+
+  if( segData.synapses.empty())
+    return;   // No synapses to work with, no work to do.
+
+  // Sort the potential pool by permanence values, and look for the synapse with
+  // the N'th greatest permanence, where N is the desired number of connected
+  // synapses.  Then calculate how much to change the N'th synapses permance by
+  // such that it becomes a connected synapse.  After that there will be exactly
+  // N synapses connected.
+  SynapseIdx N;
+  if( segData.numConnected < minimumSynapses ) {
+    N = minimumSynapses;
+  }
+  else if( segData.numConnected > maximumSynapses ) {
+    N = maximumSynapses;
+  }
+  else {
+    return;  // The segment already satisfies the requirements, done.
+  }
+  // Can't connect more synapses than there are in the potential pool.
+  N = std::min( N, (SynapseIdx) segData.synapses.size() );
+  // The N'th synapse is at index N-1
+  if( N != 0 ) {
+    N--;
+  }
+  // else {
+  //   Corner case: the user has requested a maximum of 0 connected synapses...
+  // }
+
+  vector<Permanence> permanences; permanences.reserve( segData.synapses.size() );
+  for( Synapse syn : segData.synapses )
+    permanences.push_back( synapses_[syn].permanence );
+
+  // Do a partial sort, it's faster than a full sort.
+  auto minPermPtr = permanences.begin() + (segData.synapses.size() - 1 - N);
+  std::nth_element(permanences.begin(), minPermPtr, permanences.end());
+
+  Real delta = connectedThreshold_ - *minPermPtr;
+
+  // Corner case: disconnect all synapses.
+  if( maximumSynapses == 0u ) {
+    delta -= htm::Epsilon;
```

Comment:
why? Rather 2*Epsilon, so fabs(new perm - epsilon) > 0 ?

Reply:
I'm actually going to change the assert statements to make sure this corner case never happens.